File: bench/lcg.c
#include <stdio.h>
#include <stdint.h>
#include <time.h>

static uint64_t compute(void) __attribute__((noinline));

static uint64_t compute(void) {
    uint64_t x = 1;
    for (long i = 0; i < 100000000L; i++) {
        x = x * 6364136223846793005ULL + 1442695040888963407ULL;
    }
    return x;
}

int main(void) {
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    uint64_t x = compute();
    clock_gettime(CLOCK_MONOTONIC, &t1);

    long long micros = (long long)(t1.tv_sec - t0.tv_sec) * 1000000LL
                     + (long long)(t1.tv_nsec - t0.tv_nsec) / 1000LL;

    printf("RESULT %lld\n", (long long)x);
    printf("MICROS %lld\n", micros);
    return 0;
}

File: bench/lcg.py
import time

MASK = 0xFFFFFFFFFFFFFFFF


def compute():
    x = 1
    for _ in range(100000000):
        x = (x * 6364136223846793005 + 1442695040888963407) & MASK
    return x


t0 = time.perf_counter()
x = compute()
t1 = time.perf_counter()

if x >= 2 ** 63:
    x -= 2 ** 64

micros = round((t1 - t0) * 1e6)
print("RESULT %d" % x)
print("MICROS %d" % micros)
